### ISSUE TITLE
JSONSchema Deserializer does not work with kotlin data classes

### ISSUE THREAD

**Issue by bastman** [apache/pulsar#6528]
* producer , consumer example*

The consumer fails with error: 
```
Exception in thread "main" org.apache.pulsar.client.api.SchemaSerializationException: org.apache.pulsar.shade.com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `example.simple.SensorReading` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: (byte[])"{"temperature":1.5}"; line: 1, column: 2]
	at org.apache.pulsar.client.impl.schema.reader.JsonReader.read(JsonReader.java:44)
	at org.apache.pulsar.client.api.schema.SchemaReader.read(SchemaReader.java:36)
	at org.apache.pulsar.client.impl.schema.StructSchema.decode(StructSchema.java:92)
	at org.apache.pulsar.client.impl.MessageImpl.getValue(MessageImpl.java:278)
	at example.simple.SimpleAppKt.main(SimpleApp.kt:52)
	at example.simple.SimpleAppKt.main(SimpleApp.kt)
Caused by: org.apache.pulsar.shade.com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `example.simple.SensorReading` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: (byte[])"{"temperature":1.5}"; line: 1, column: 2]
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1589)
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1055)
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1297)
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326)
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159)
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4202)
	at org.apache.pulsar.shade.com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3275)
	at org.apache.pulsar.client.impl.schema.reader.JsonReader.read(JsonReader.java:42)

```

```
package example.simple

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.apache.pulsar.client.api.*
import org.apache.pulsar.client.impl.schema.JSONSchema


private const val SERVICE_URL = "pulsar://localhost:6650"
private const val TOPIC_NAME = "test-topic-jsonschema-002"
private const val SUBSCRIPTION_NAME = "test-subscription-001"
private typealias PulsarProducer = Producer<SensorReading>
private typealias PulsarConsumer = Consumer<SensorReading>

data class SensorReading(val temperature: Float)

fun main() {
    val schema = JSONSchema.of(SensorReading::class.java)
    val client: PulsarClient by lazy {
        PulsarClient.builder()
                .serviceUrl(SERVICE_URL)
                .build()
    }
    val producer: PulsarProducer = client.newProducer(schema)
            .topic(TOPIC_NAME)
            .compressionType(CompressionType.LZ4)
            .create()

    val consumer: PulsarConsumer = client.newConsumer(schema)
            .topic(TOPIC_NAME)
            .subscriptionType(SubscriptionType.Shared)
            .subscriptionName(SUBSCRIPTION_NAME)
            .subscribe()

    // check json (stringify/parse) works
    val JSON = jacksonObjectMapper()
    val source = SensorReading(temperature = 2.5f)
    val txt = JSON.writeValueAsString(source)
    val sink: SensorReading = JSON.readValue(txt)

    run {
        val content = SensorReading(temperature = 1.5f)
        val msg = producer.newMessage()
                .value(content)
        msg.send()
    }

    run {
        while (true) {
            val msg = consumer.receive()
            println("received ${msg.messageId}")
            val content = msg.value
            println("Message received: $content")

            // Acknowledge the message so that it can be deleted by the message broker
            consumer.acknowledge(msg)
            println("Ack sent.")
        }
    }


}


```

* side notes *
jackson-module-kotlin is not used by the build in objectmapper

Is there any chance to get that supported?
alternative: let's pass in the pre-configured objectmapper as argument of schema factory.

**Comment by kindofwhat** [comment]
as a workaround: supply default values for all properties. this of course is not "proper" support for data classes,,,

**Comment by tomas-c** [comment]
Since #6905 has been merged, it's now possible to create a schema that uses a custom ObjectMapper like so:

```
val objectMapper= ObjectMapper()
objectMapper.registerModule(KotlinModule())
objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL)

val schemDefinition = SchemaDefinition.builder<YourClass>()
            .withPojo(YourClass::class.java)
            .withSchemaReader(JacksonJsonReader(objectMapper, YourClass::class.java))
            .withSchemaWriter(JacksonJsonWriter(objectMapper))
            .build()
```

Note that this only works with the unshaded version of the library (use org.apache.pulsar:pulsar-client-original instead of org.apache.pulsar:pulsar-client)

**Comment by sijie** [comment]
@tomas-c Thank you for calling it now.

Close the issue since it is fixed by #6905
